zerver/tests/test_subs.py

```diff
@@ -2198,6 +2204,59 @@ def test_unarchive_stream(self) -> None:
             },
         )
 
+    def test_permissions_and_archiving_behavior(self) -> None:
+        """
+        Test permissions for archiving and unarchiving streams, and ensure users without
+        the necessary permissions cannot archive a stream.
+        """
+        hamlet = self.example_user("hamlet")
+        cordelia = self.example_user("cordelia")
+        self.login_user(hamlet)
+
+        stream = self.make_stream("test_stream", invite_only=False)
+        self.subscribe(hamlet, stream.name)
+        do_change_user_role(hamlet, UserProfile.ROLE_REALM_ADMINISTRATOR, acting_user=None)
+
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_success(result)
+
+        self.login_user(cordelia)
+        result = self.client_patch(
+            f"/json/streams/{stream.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+        self.login_user(hamlet)
+        result = self.client_patch(
+            f"/json/streams/{stream.id}", {"is_archived": orjson.dumps(False).decode()}
+        )
+        self.assert_json_success(result)
+
+        self.login_user(cordelia)
+        result = self.client_delete(f"/json/streams/{stream.id}")
+        self.assert_json_error(result, "You do not have permission to administer this channel.")
+
+        stream.refresh_from_db()
+        self.assertFalse(stream.deactivated)
```

Comment:
Would be good to test the case for non-admin users having permission through `can_administer_channel_group`.